libs/constrained-generators/src/Constrained/Base.hs

```diff
@@ -3868,55 +3877,138 @@ instance HasSpec fn a => HasSpec fn [a] where
       genError1 "combineSpec ListSpec failed with <REASON>"
     when badSizeSpec $
       genError1 "error-like size spec in combineSpec ListSpec"
-    typeSpec . ListSpec (unionWithMaybe min msz msz') must'' size'' elemS''
-      <$> combineFoldSpec foldS foldS'
-
-  genFromTypeSpec (ListSpec _ must _ elemS _)
-    | any (not . (`conformsToSpec` elemS)) must =
-        genError1 "genTypeSpecSpec @ListSpec: must do not conform to elemS"
-  genFromTypeSpec (ListSpec msz must TrueSpec elemS NoFold) = do
+    fold'' <- combineFoldSpec foldS foldS'
+    split'' <- combineSplit sp sp'
+    pure $ typeSpec (ListSpec (unionWithMaybe min msz msz') must'' size'' elemS'' fold'' split'')
+
+  genFromTypeSpec (ListSpec _ must _ elemS _ split)
+    | any (not . (`conformsToSpec` elemS)) (must ++ splitContents split) =
+        genError1 "genTypeSpecSpec @ListSpec: (must set) or (split contents) do not conform to elemS"
+  genFromTypeSpec (ListSpec _ must size _ NoFold (Split (Just xs) (Just ys)))
+    | not (conformsToSpec (sizeOf (xs ++ ys)) size) = genError1 "Size and split don't agree"
+    | not (all (`elem` (xs ++ ys)) must) =
```

Comment:
This is an O(m*n) operation. Are you sure at least one of them is always small? Otherwise it will be very expensive!